src/models/userHelpers.js

```diff
@@ -627,25 +643,71 @@ async function getExistingUsers(d2, options = {}) {
     return users;
 }
 
-function getPayload(parentUser, destUsers) {
-    const users = destUsers.map(childUser => {
-        let childUserRoles = childUser.userCredentials.userRoles;
-        if (childUserRoles.length == 0) {
-            return _m.imerge(childUser, {
-                userCredentials: _m.imerge(childUser.userCredentials, {
-                    userRoles: parentUser.userCredentials.userRoles,
-                }),
-            });
-        } else {
-            parentUser.userCredentials.userRoles.forEach(role => {
-                if (childUserRoles.find(element => element.id == role.id) == undefined) {
-                    childUserRoles.push(role);
+/*
+use cases:
+1 parent to 1 child:
+    1. userRoles :) 
+    2. userGroups :) 
+    3. userRoles AND userGroups :)
+    4. child has no userRoles :)
+    5. child has no userGroups :)
+    5. child has no userRoles AND userGroups :)
+
+1 parent to many children
+    1. userRoles
+    2. userGroups
+    3. userRoles AND userGroups :)
+    4. child has no userRoles 
+    5. child has no userGroups 
+    5. child has no userRoles AND userGroups 
+*/
+function getPayload(parentUser, destUsers, copyUserGroups, copyUserRoles) {
+    
+        const users = destUsers.map(childUser => {
+            let childUserRoles = childUser.userCredentials.userRoles;
+            let childUserGroups = childUser.userGroups;
+            if(childUserGroups.length == 0 && childUserRoles.length == 0)
+            {
+                return _m.imerge(childUser, {
+                    userCredentials: _m.imerge(childUser.userCredentials, {
+                        userRoles: parentUser.userCredentials.userRoles,
+                    }),
+                    userGroups: parentUser.userGroups,
+                });  
+            }
+            if(copyUserRoles)
+            {   
+                if (childUserRoles.length == 0) {
+                    return _m.imerge(childUser, {
```

Comment:
I don't see why this `return` here, it stops the iteration when there `copyUserGroups` may be true.

As far as I can see, the destination array (roles or groups) being empty is not a special case, just push the values from parentUser, as you do in the next block.